.github/workflows/matrix.json

```
@@ -0,0 +1,34 @@
+{
+  "python-version": [
+    "3.9",
+    "3.10",
+    "3.11",
+    "3.12"
+  ],
+  "singularity": [
+      {
+        "provider": "apptainer",
+        "version": "1.3.4"
+      },
+      {
+        "provider": "singularity-ce",
+        "version": "4.3.0"
+      },
+      {
+        "provider": "singularity-ce",
+        "version": "4.2.1"
+      }
+  ],
+  "os":[
+    {
+     "runs_on": "ubuntu-latest",
```

Maybe
```suggestion
     "runs-on": "ubuntu-latest",
```
for consistency

---

separate push